File: 5_ABSTRACT_DATA/Java/Stack_Parametrico/Stack.java
class Stack<Type> {

	// Properties
	private Object[] stackStorage;
	private int stackTopPointer;

	// Methods
	
	public Stack(int myMaxStackDim) {
		stackStorage = new Object[myMaxStackDim];
		stackTopPointer = 0;
	}

	public void push(Type elem) {
		if (stackTopPointer < stackStorage.length)
			stackStorage[stackTopPointer++] = elem;
	}

	public void pop() {
		if (stackTopPointer > 0) stackTopPointer--;
	}

	@SuppressWarnings("unchecked")
	public Type top() {
		if (stackTopPointer > 0) 
			return (Type) stackStorage[stackTopPointer - 1];
		else
			return null;
	}

	public boolean empty() {
		return (stackTopPointer == 0);
	}
}

File: 5_ABSTRACT_DATA/C/stack.c
#include <stdlib.h>
#include "stack.h"

void create(Stack * s) {
	(*s).v = (element *) malloc (MAX_STACK_SIZE * sizeof(element) );
	(*s).p = 0;
}

void destroy(Stack * s) {
	free( (*s).v );
	(*s).p = -1;
}

int empty(Stack s) {
	return ( s.p == 0 );
}

void push(Stack * s, element elem) {
	if ( (*s).p < MAX_STACK_SIZE )
		(*s).v[ (*s).p++ ] = elem;
}

void pop(Stack * s) {
	if ( (*s).p > 0 )
		(*s).p--;
}

element top(Stack s) {
	if ( s.p > 0 )
		return s.v[ s.p - 1 ];
	else
		return -1;
}
